// loopback4-notifications/dist/providers/sms/sns/sns.provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnsProvider = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@loopback/core");
const rest_1 = require("@loopback/rest");
const aws_sdk_1 = require("aws-sdk");
const keys_1 = require("./keys");
let SnsProvider = class SnsProvider {
    constructor(snsConfig) {
        this.snsConfig = snsConfig;
        if (this.snsConfig) {
            this.snsService = new aws_sdk_1.SNS(this.snsConfig);
        }
        else {
            throw new rest_1.HttpErrors.PreconditionFailed('AWS SNS Config missing !');
        }
    }
    value() {
        return {
            publish: async (message) => {
                if (message.receiver.to.length === 0) {
                    throw new rest_1.HttpErrors.BadRequest('Message receiver not found in request');
                }
                const publishes = message.receiver.to.map(receiver => {
                    var _a;
                    const msg = {
                        Message: message.body,
                        Subject: message.subject,
                    };
                    if ((_a = message.options) === null || _a === void 0 ? void 0 : _a.smsType) {
                        msg.MessageAttributes = {
                            'AWS.SNS.SMS.SMSType': {
                                DataType: 'String',
                                StringValue: message.options.smsType,
                            },
                        };
                    }
                    if (receiver.type === 0 /* PhoneNumber */) {
                        msg.PhoneNumber = receiver.id;
                    }
                    else if (receiver.type === 1 /* Topic */) {
                        msg.TopicArn = receiver.id;
                    }
                    else {
                        // Do nothing
                    }
                    return this.snsService.publish(msg).promise();
                });
                await Promise.all(publishes);
            },
        };
    }
};
SnsProvider = (0, tslib_1.__decorate)([
    (0, tslib_1.__param)(0, (0, core_1.inject)(keys_1.SNSBindings.Config, {
        optional: true,
    })),
    (0, tslib_1.__metadata)("design:paramtypes", [Object])
], SnsProvider);
exports.SnsProvider = SnsProvider;
